pkg/pillar/types/global.go

```diff
@@ -103,312 +105,423 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+	specMap.GlobalSettings[key] = configItem
+	log.Debugf("Added string item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) AddTriStateItem(key GlobalSettingKey, defaultTriState TriState) {
+	configItem := ConfigItemSpec{
+		Key:             string(key),
+		ItemType:        ConfigItemTypeTristate,
+		TriStateDefault: defaultTriState,
 	}
-	if newgc.MetricInterval < GlobalConfigMinimums.MetricInterval {
-		log.Warnf("Enforce minimum MetricInterval received %d; using %d",
-			newgc.MetricInterval, GlobalConfigMinimums.MetricInterval)
-		newgc.MetricInterval = GlobalConfigMinimums.MetricInterval
+	specMap.GlobalSettings[key] = configItem
+	log.Debugf("Added tristate item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) AddAgentSettingStringItem(key AgentSettingKey, defaultString string) {
+	configItem := ConfigItemSpec{
+		ItemType:      ConfigItemTypeString,
+		Key:           string(key),
+		StringDefault: defaultString,
 	}
-	if newgc.ResetIfCloudGoneTime < GlobalConfigMinimums.ResetIfCloudGoneTime {
-		log.Warnf("Enforce minimum XXX received %d; using %d",
-			newgc.ResetIfCloudGoneTime, GlobalConfigMinimums.ResetIfCloudGoneTime)
-		newgc.ResetIfCloudGoneTime = GlobalConfigMinimums.ResetIfCloudGoneTime
+	specMap.AgentSettings[key] = configItem
+	log.Debugf("Added string item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) ParseItem(configMap *ConfigItemValueMap, value string, key string) error {
+	// legacy per-agent setting key debug.<agentname>.xxx
+	// new per-agent setting key agent.<agentname>.debug.xxx
+	var itemSpec ConfigItemSpec
+	agentSetting := false
+	globalSetting := false
+	var agentName string
+	components := strings.Split(key, ".")
+	if strings.HasPrefix(key, "agent") || specMap.isLegacyAgent(key) {
+		agentName = components[1]
+		if strings.HasPrefix(key, "agent") && len(components) > 2 {
+			components = components[2:]
+		} else if strings.HasPrefix(key, "debug") && len(components) > 2 {
+			components[1] = ""
+		} else {
+			return fmt.Errorf("Unable to find agent name for per-agent setting. Key: %s", key)
+		}
+		key = strings.Join(components, ".")
+		agentSetting = true
+	} else if _, ok := specMap.GlobalSettings[GlobalSettingKey(key)]; ok {
+		globalSetting = true
+	} else {
+		return fmt.Errorf("Item is neither a global nor a per-agent setting. Key: %s", key)
+	}
+	if agentSetting {
+		itemSpec = specMap.AgentSettings[AgentSettingKey(key)]
+		val, err := itemSpec.parseValue(value)
+		if err == nil {
+			agentMap, ok := configMap.AgentSettings[agentName]
+			if !ok {
+				agentMap = make(map[AgentSettingKey]ConfigItemValue)
+			}
+			agentMap[AgentSettingKey(key)] = val
+			configMap.AgentSettings[agentName] = agentMap
+		} else {
+			return err
+		}
+	} else if globalSetting {
+		itemSpec = specMap.GlobalSettings[GlobalSettingKey(key)]
+		val, err := itemSpec.parseValue(value)
+		if err == nil {
+			configMap.GlobalSettings[GlobalSettingKey(key)] = val
+		} else {
+			return err
+		}
+	}
```

Comment:
Same here.. why not just put this in the above if block:

else if _, ok := specMap.GlobalSettings[GlobalSettingKey(key)]; ok {
		globalSetting = true